Doc review of Global Configuration doc (#1223)
SAP/spartacus-docs: _pages/dev/global-configuration-in-spartacus.md

title: Global Configuration in Spartacus
---

Spartacus uses a mechanism that provides global configuration during app initialization (that is, when the application is bootstrapped). This configuration does not change while the application is running. Each storefront module that uses this configuration usually provides typing with some defaults for its part of the configuration.

> If necessary, you can always override the global configuration in the main app module.

Author: Ah, I see! Based on your comment, I am changing the line into a note:

>Note: The configuration in the main app module takes precedence over other configurations, and can be used to override any configuration that has been provided elsewhere.

> If necessary, you can always override the global configuration in the main app module.

Reviewer: 👍

***

## Providing Global Configurations in Spartacus

The following sections describe the different ways you can provide global configuration in Spartacus.

### StorefrontLib.withConfig

If you import the `B2cStorefrontModule` in your app, the preferred and easiest method to provide a configuration is to use `B2cStorefrontModule.withConfig(config?: StorefrontModuleConfig)`. The `StorefrontModuleConfig` interface provides type safety and context-aware code completion that speeds up defining the configuration and helps to avoid any typos.

### ConfigModule.withConfig

Importing `ConfigModule.withConfig(config: any)` is useful when you want to use the configuration in your module and contribute to it at the same time. The `config` parameter is not typed and can be an object of any shape, so it is recommended that you use either a typed `const`, or that you use `ConfigModule.withConfig(<ConfigType>{})` type casting to be able to use type safety.

> ### provideConfig

Reviewer: ProvideConfig should be used as preffered one, `ConfigModule.withConfig` as the legacy one, and `(B2c)StorefrontModule.withConfig` are dropped already.

Author: I have changed the order of the sections to reflect your comment, I have pretty much removed the contents of the `StorefrontModule.withConfig` section (except to say that it is no longer supported), and I have added a line at the beginning of the other sections to say which one is the preferred method and which one is the legacy method.

Beyond that, I assume we should modify what the text currently says about `provideConfig` and `ConfigModule.withConfig`. Do you have any suggestions? Or is there somone I can follow up with to ask for a "rewrite" of these sections? 🙂

Reviewer: 👍

***

## Providing Global Configurations in Spartacus

The following sections describe the different ways you can provide global configuration in Spartacus.

### provideConfig

Using `provideConfig` is the preferred way for providing global configuration in Spartacus.

> If you want to contribute to the global configuration without needing to import the `ConfigModule`, or you want to implement a module with providers and provide the configuration conditionally, then using `provideConfig` in a providers array is the best option.

Reviewer: ```suggestion
If you want to contribute to the global configuration using `provideConfig`, just add it to a providers array in an Angular module.
```

### ConfigModule.withConfig

Using `ConfigModule.withConfig` is the legacy method of providing global configuration in Spartacus.

> Importing `ConfigModule.withConfig(config: any)` is useful when you want to use the configuration in your module and contribute to it at the same time. The `config` parameter is not typed and can be an object of any shape, so it is recommended that you use either a typed `const`, or that you use `ConfigModule.withConfig(<ConfigType>{})` type casting to be able to use type safety.

Reviewer: ```suggestion
Importing `ConfigModule.withConfig(config: Config)` is useful when you want to use the configuration in your module and contribute to it at the same time. 
```

## Modifying the Configuration at Runtime

The global configuration mechanism in Spartacus is not designed to modify the configuration after the application has started. If you need to modify the configuration at runtime (that is, after bootstrap), you can put the configuration in a store, or expose it in a service, ideally as an observable stream that can easily react to changes.

However, the general recommendation is that any value that can change over time should be part of the application state, rather than part of the application configuration.

## Default Values

Each module typically provides a default configuration that is required for basic functionality. A default configuration is any configuration that is provided inside the related module. It usually contains reasonable defaults or configurations that are required for a module to operate.

You do not need to provide defaults for all required configurations. For example, it may be difficult to choose reasonable defaults for some options. In this case, it is recommended to use config validators to validate the configuration and warn (in development mode) if a required config is missing.

> **Note:** Default configurations are provided in the same way as any other configuration.

Author: Note changed to reflect this info 🙂

> **Note:** Default configurations are provided in the same way as any other configuration.

Reviewer: 👍
